ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/BlockingQueryPublisher.java

```diff
@@ -59,13 +60,23 @@ public BlockingQueryPublisher(final Context ctx,
     this.workerExecutor = Objects.requireNonNull(workerExecutor);
   }
 
-  public void setQueryHandle(final PushQueryHandle queryHandle) {
+  public void setQueryHandle(final QueryHandle queryHandle, final boolean isPullQuery) {
     this.columnNames = queryHandle.getColumnNames();
     this.columnTypes = queryHandle.getColumnTypes();
     this.queue = queryHandle.getQueue();
+    this.isPullQuery = isPullQuery;
     this.queue.setQueuedCallback(this::maybeSend);
-    this.queue.setLimitHandler(() -> complete = true);
+    this.queue.setLimitHandler(() -> {
+      complete = true;
+      // This allows us to hit the limit without having to queue one last row
+      if (queue.isEmpty()) {
+        ctx.runOnContext(v -> sendComplete());
+      } else {
+        ctx.runOnContext(v -> doSend());
+      }
```

Comment:
I'm confused. Why is this necessary? Why would we otherwise have to enqueue another row?